garage/envs/base.py

```diff
@@ -1,17 +1,121 @@
+"""Wrapper class that converts gym.Env into GarageEnv."""
+from abc import ABC, abstractmethod
 import collections
+import warnings
 
-from cached_property import cached_property
+import gym
 
-from garage.envs import EnvSpec
+from garage.core import Parameterized
+from garage.core import Serializable
 
-_Step = collections.namedtuple("Step",
-                               ["observation", "reward", "done", "info"])
 
+class GarageEnv(ABC, gym.Wrapper, Parameterized, Serializable):
+    """
+    Returns an abstract Garage wrapper class for gym.Env.
+
+    In order to provide pickling (serialization) and parameterization
+    for gym.Envs, they must be wrapped with a GarageEnv. This ensures
+    compatibility with existing samplers and checkpointing when the
+    envs are passed internally around garage.
 
-def Step(observation, reward, done, **kwargs):
+    Furthermore, classes inheriting from GarageEnv should silently
+    convert action_space and observation_space from gym.Spaces to
+    garage.spaces.
+
+    Args: env (gym.Env): the env that will be wrapped
     """
-    Convenience method creating a namedtuple with the results of the
-    environment.step method.
-    Put extra diagnostic info in the kwargs
+
+    def __init__(self, env):
+        Serializable.quick_init(self, locals())
+        Parameterized.__init__(self)
+        super().__init__(env)
+
+    def close(self):
+        """
+        Close the wrapped env.
+
+        Returns:
+            None
+        """
+        self.env.close()
+
+    def get_params_internal(self, **tags):
+        """
+        Returns an empty list if env.get_params() is called.
+
+        Returns:
+            An empty list
+        """
+        warnings.warn("get_params_internal is deprecated", DeprecationWarning)
+        return []
+
+    @property
+    def horizon(self):
+        """
+        Get the maximum episode steps for the wrapped env.
+
+        Returns:
+            max_episode_steps (int)
+        """
+        if self.env.spec is not None:
+            return self.env.spec.max_episode_steps
+        else:
+            return NotImplementedError
+
+    def log_diagnostics(self, paths, *args, **kwargs):
+        """No env supports this function call."""
+        warnings.warn("log_diagnostics is deprecated", DeprecationWarning)
+        pass
+
+    @abstractmethod
+    def spec(self):
+        """
+        Returns an EnvSpec with garage.spaces.
+
+        Returns:
+            spec (garage.envs.EnvSpec)
+        """
+        raise NotImplementedError
+
+    def reset(self, **kwargs):
+        """
+        Inheriting gym.Wrapper requires implementing this method.
+
+        Calls reset on wrapped env.
+        """
+        return self.env.reset(**kwargs)
```

Comment:
most of these should be provided by gym.Wrapper. Are you sure the gym.Wrapper constructor is being called?

Reply:
Yes, gym.Wrapper requires all wrappers derived from gym.Wrapper to implement step() and reset(), or else it throws a deprecated warning:
```
    def step(self, action):
        if hasattr(self, "_step"):
            deprecated_warn_once("%s doesn't implement 'step' method, but it implements deprecated '_step' method." % type(self))
            self.step = self._step
            return self.step(action)
        else:
            deprecated_warn_once("%s doesn't implement 'step' method, " % type(self) +
                "which is required for wrappers derived directly from Wrapper. Deprecated default implementation is used.")
            return self.env.step(action)

    def reset(self, **kwargs):
        if hasattr(self, "_reset"):
            deprecated_warn_once("%s doesn't implement 'reset' method, but it implements deprecated '_reset' method." % type(self))
            self.reset = self._reset
            return self._reset(**kwargs)
        else:
            deprecated_warn_once("%s doesn't implement 'reset' method, " % type(self) +
                "which is required for wrappers derived directly from Wrapper. Deprecated default implementation is used.")
            return self.env.reset(**kwargs)
```